pootle/apps/pootle_project/lang_mapper.py

```diff
@@ -0,0 +1,96 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+import logging
+from collections import OrderedDict
+
+from django.utils.functional import cached_property
+from django.utils.lru_cache import lru_cache
+
+from pootle.core.delegate import config
+from pootle_language.models import Language
+
+
+logger = logging.getLogger(__name__)
+
+
+class LanguageMapper(object):
+    """For a given code find the relevant pootle lang taking account of any
+    lang mapping configuration
+
+    """
+
+    def __init__(self, project):
+        self.project = project
+
+    def __getitem__(self, k):
+        return self.get_lang(k)
+
+    def __contains__(self, k):
+        return (
+            k in self.lang_mappings
+            and self.get_lang(k))
+
+    @property
+    def project_mapping(self):
+        return config.get(
+            self.project.__class__,
+            instance=self.project,
+            key="pootle.core.lang_mapping") or {}
+
+    @property
+    def site_presets(self):
+        return config.get(
+            key="pootle.core.lang_mapping_presets") or {}
+
+    @cached_property
+    def lang_mappings(self):
+        return self._parse_mappings()
+
+    def get_upstream_code(self, pootle_code):
+        for fs_code, pootle_code in self.lang_mappings.items():
+            if pootle_code == pootle_code:
+                return fs_code
+        return pootle_code
+
+    @lru_cache()
+    def get_lang(self, lang_code):
+        try:
+            return Language.objects.get(
+                code=self.get_pootle_code(lang_code))
+        except Language.DoesNotExist:
+            return None
+
+    def get_pootle_code(self, lang_code):
+        return self.lang_mappings.get(lang_code, lang_code)
+
+    def _add_lang_to_mapping(self, k, v):
+        # as its a 1-2-1 mapping remove any previous items with
+        # same value
+        if v in self._mapping.values():
+            for _k, _v in self._mapping.items():
+                if v == _v:
+                    del self._mapping[_k]
+                    break
+        self._mapping[k] = v
+
+    def _parse_mappings(self):
+        self._mapping = OrderedDict()
+
+        for k, v in self.project_mapping.items():
```

Comment:
`k` and `v` do not convey any information. But I see that above `fs_code` and `pootle_code` are used, and those are easier to understand.